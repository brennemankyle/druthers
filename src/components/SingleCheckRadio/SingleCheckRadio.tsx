import React, { ReactElement, useState, MouseEventHandler } from "react";
import { noop } from "../../utils/utils";
import { MassagedSelectProps } from "../../utils/SelectTypes";

interface Props {
  checked: boolean;
  toggle: boolean;
  value: string;
  onClick: MouseEventHandler<HTMLInputElement>;
  className: string;
  type: string;
  label: ReactElement | string;
  title: string;
}

type CheckRadioProps = Partial<Props> & MassagedSelectProps;

function SingleCheckRadio({
  onClick = noop,
  multiple = false,
  label: baseLabel = "",
  title = "",
  toggle,
  type: baseType,
  className: baseClassName,
  svg_Checkmark,
  value,
  name,
  onChange,
  disabled,
  checked,
}: CheckRadioProps): ReactElement {
  let [hasFocus, setHasFocus] = useState(false);

  let onFocus = () => setHasFocus(true);
  let onBlur = () => setHasFocus(false);
  let type = toggle
    ? "checkbox"
    : baseType || (multiple ? "checkbox" : "radio");
  let label = baseLabel || title;
  let className = baseClassName + " " + (hasFocus ? "focus" : ""); // Allow styling based on .focus
  let Checkmark = svg_Checkmark;

  return (
    <label className={className} key={value}>
      <input
        type={type}
        name={name}
        value={value}
        onChange={onChange}
        onClick={onClick}
        disabled={disabled}
        onFocus={onFocus}
        onBlur={onBlur}
        checked={checked}
      />
      <div className="check-radio-display">
        <Checkmark />
      </div>
      {label}
    </label>
  );
}

export default SingleCheckRadio;
