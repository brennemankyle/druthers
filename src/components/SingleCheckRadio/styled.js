import styled from "@emotion/styled";
import SingleCheckRadio from "./SingleCheckRadio";

const CheckBox = styled(SingleCheckRadio)`
  margin-right: ${props => props.styles_checkRadio_marginBetween};
  display: inline-flex;
  align-items: center;
  user-select: none;
  cursor: ${props => (props.disabled ? "not-allowed" : "pointer")};
  ${props =>
    props.disabled
      ? `&& .check-radio-display {
        border-color: ${props.styles_colors_disabled};
      }`
      : `&:hover .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }
      &.focus .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }`}

  .check-radio-display {
    border-radius: ${props => props.styles_borderRadius};
    margin-right: ${props => props.styles_checkRadio_labelMargin};
    padding-top: ${props => props.styles_checkRadio_paddingTop};
    padding-bottom: ${props => props.styles_checkRadio_paddingBottom};
    padding-right: ${props => props.styles_checkRadio_paddingRight};
    padding-left: ${props => props.styles_checkRadio_paddingLeft};
    border: ${props => props.styles_checkRadio_borderWidth} solid
      ${props => props.styles_colors_secondary};
    width: ${props => props.styles_icon_width};
    height: ${props => props.styles_icon_width};
    background-color: ${props =>
      props.disabled
        ? props.styles_colors_disabled
        : props.styles_colors_background};

    svg {
      fill: ${props => props.styles_colors_primary};
      width: ${props => props.styles_icon_width};
      height: ${props => props.styles_icon_width};
      ${props => !props.checked && `visibility: hidden;`}
    }
  }

  input {
    opacity: 0;
    width: 0;
  }
`;

const Radio = styled(SingleCheckRadio)`
  margin-right: ${props => props.styles_checkRadio_marginBetween};
  display: inline-flex;
  align-items: center;
  cursor: pointer;
  user-select: none;
  cursor: ${props => (props.disabled ? "not-allowed" : "pointer")};
  ${props =>
    props.disabled
      ? `&& .check-radio-display {
        border-color: ${props.styles_colors_disabled};
      }`
      : `&:hover .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }
      &.focus .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }`}

  .check-radio-display {
    border-radius: 100%;
    margin-right: ${props => props.styles_checkRadio_labelMargin};
    padding-top: ${props => props.styles_checkRadio_paddingTop};
    padding-bottom: ${props => props.styles_checkRadio_paddingBottom};
    padding-right: ${props => props.styles_checkRadio_paddingRight};
    padding-left: ${props => props.styles_checkRadio_paddingLeft};
    background-size: 0.1em;
    border: ${props => props.styles_checkRadio_borderWidth} solid
      ${props => props.styles_colors_secondary};
    width: ${props => props.styles_icon_width};
    height: ${props => props.styles_icon_width};
    background-color: ${props =>
      props.disabled
        ? props.styles_colors_disabled
        : props.styles_colors_background};
    ${props =>
      props.checked && `background-color: ${props.styles_colors_secondary};`}

    svg {
      visibility: hidden;
      width: ${props => props.styles_icon_width};
      height: ${props => props.styles_icon_width};
    }
  }

  input {
    opacity: 0;
    width: 0;
  }
`;

const Switch = styled(SingleCheckRadio)`
  display: inline-flex;
  align-items: center;
  cursor: pointer;
  user-select: none;
  cursor: ${props => (props.disabled ? "not-allowed" : "pointer")};
  ${props =>
    props.disabled
      ? `&& .check-radio-display {
        border-color: ${props.styles_colors_disabled};
      }`
      : `&:hover .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }
      &.focus .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }`}

  .check-radio-display {
    margin-right: ${props => props.styles_checkRadio_labelMargin};
    display: inline-flex;
    align-items: center;
    border-radius: 1em;
    padding-top: ${props => props.styles_checkRadio_paddingTop};
    padding-bottom: ${props => props.styles_checkRadio_paddingBottom};
    padding-right: ${props => props.styles_checkRadio_paddingRight};
    padding-left: ${props => props.styles_checkRadio_paddingLeft};
    border: ${props => props.styles_checkRadio_borderWidth} solid
      ${props => props.styles_colors_secondary};
    width: calc(${props => props.styles_icon_width} * 2);
    height: ${props => props.styles_icon_width};
    background-color: ${props =>
      props.disabled
        ? props.styles_colors_disabled
        : props.styles_colors_background};

    svg {
      width: ${props => props.styles_icon_width};
      height: ${props => props.styles_icon_width};
      border-radius: 100%;
      ${props =>
        props.checked
          ? `margin-left: ${props.styles_icon_width};
          background-color: ${props.styles_colors_primary};`
          : `background-color: ${props.styles_colors_secondary};`}
      fill-opacity: 0;
    }
  }

  input {
    opacity: 0;
    width: 0;
  }
`;

export { CheckBox, Radio, Switch };
