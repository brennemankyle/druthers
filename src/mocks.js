let mockStyles = {
  colors: {}
}

let mockItem = {
  value: '42',
  label: 'Meaning of life',
}

let mockItemList = [
  {value: '1', label: 'Item 1'},
  {value: '2', label: 'Item 2'},
  {value: '3', label: 'Item 3'},
  {value: '4', label: 'Item 4'},
  {value: '5', label: 'Item 5'},
]

export { mockStyles, mockItem, mockItemList }
