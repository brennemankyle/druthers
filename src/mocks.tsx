import React from "react";
import defaultProps from "./utils/defaultProps";
import { withKeys, noop } from "./utils/utils";

let mockStyles = {
  styles_width: 600,
  styles_multiple: false,
  styles_disabled: false,
  styles_hasSelection: false,
  styles_hasOptions: true,
  styles_optionHighlighted: [],
  styles_rightToLeft: false,
  ...withKeys(defaultProps, "styles_"),
};

let mockItem = {
  value: "42",
  label: "Meaning of life",
};

let mockItemList = [
  { value: "1", label: "Item 1" },
  { value: "2", label: "Item 2" },
  { value: "3", label: "Item 3" },
  { value: "4", label: "Item 4" },
  { value: "5", label: "Item 5" },
];

let mockEvent = {
  preventDefault: noop,
};

let MockElement = (props: any) => <div data-props={props} />;
let MockInput = (props: any) => <input type="text" data-props={props} />;
let MockSvg = () => (
  <svg>
    <path d="M0 0L5 5Z" />
  </svg>
);

let mockRect = {
  top: 10,
  bottom: 110,
  left: 20,
  right: 220,
  width: 200,
  height: 100,
  x: 15,
  y: 25,
};

export {
  mockStyles,
  mockItem,
  mockItemList,
  mockEvent,
  MockElement,
  MockInput,
  MockSvg,
  mockRect,
};
