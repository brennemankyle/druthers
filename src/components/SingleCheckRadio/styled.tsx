import styled from "@emotion/styled";
import SingleCheckRadio from "./SingleCheckRadio";

const CheckBox = styled(SingleCheckRadio)(
  props => `
  margin-right: ${props.styles_checkRadio_marginBetween};
  display: inline-flex;
  align-items: center;
  user-select: none;
  cursor: ${props.disabled ? "not-allowed" : "pointer"};
  ${
    props.disabled
      ? `&& .check-radio-display {
        border-color: ${props.styles_colors_disabled};
      }`
      : `&:hover .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }
      &.focus .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }`
  }

  .check-radio-display {
    border-radius: ${props.styles_borderRadius};
    margin-right: ${props.styles_checkRadio_labelMargin};
    padding-top: ${props.styles_checkRadio_paddingTop};
    padding-bottom: ${props.styles_checkRadio_paddingBottom};
    padding-right: ${props.styles_checkRadio_paddingRight};
    padding-left: ${props.styles_checkRadio_paddingLeft};
    border: ${props.styles_checkRadio_borderWidth} solid
      ${props.styles_colors_secondary};
    width: ${props.styles_icon_width};
    height: ${props.styles_icon_width};
    background-color: ${
      props.disabled
        ? props.styles_colors_disabled
        : props.styles_colors_background
    };

    svg {
      fill: ${props.styles_colors_primary};
      width: ${props.styles_icon_width};
      height: ${props.styles_icon_width};
      ${!props.checked ? `visibility: hidden;` : ``}
    }
  }

  input {
    opacity: 0;
    width: 0;
  }
`
);

const Radio = styled(SingleCheckRadio)(
  props => `
  margin-right: ${props.styles_checkRadio_marginBetween};
  display: inline-flex;
  align-items: center;
  cursor: pointer;
  user-select: none;
  cursor: ${props.disabled ? "not-allowed" : "pointer"};
  ${
    props.disabled
      ? `&& .check-radio-display {
          border-color: ${props.styles_colors_disabled};
        }`
      : `&:hover .check-radio-display {
          border-color: ${props.styles_colors_primary};
        }
        &.focus .check-radio-display {
          border-color: ${props.styles_colors_primary};
        }`
  }

  .check-radio-display {
    border-radius: 100%;
    margin-right: ${props.styles_checkRadio_labelMargin};
    padding-top: ${props.styles_checkRadio_paddingTop};
    padding-bottom: ${props.styles_checkRadio_paddingBottom};
    padding-right: ${props.styles_checkRadio_paddingRight};
    padding-left: ${props.styles_checkRadio_paddingLeft};
    background-size: 0.1em;
    border: ${props.styles_checkRadio_borderWidth} solid
      ${props.styles_colors_secondary};
    width: ${props.styles_icon_width};
    height: ${props.styles_icon_width};
    background-color: ${
      props.disabled
        ? props.styles_colors_disabled
        : props.styles_colors_background
    };
    ${
      props.checked ? `background-color: ${props.styles_colors_secondary};` : ``
    }

    svg {
      visibility: hidden;
      width: ${props.styles_icon_width};
      height: ${props.styles_icon_width};
    }
  }

  input {
    opacity: 0;
    width: 0;
  }
`
);

const Switch = styled(SingleCheckRadio)(
  props => `
  display: inline-flex;
  align-items: center;
  cursor: pointer;
  user-select: none;
  cursor: ${props.disabled ? "not-allowed" : "pointer"};
  ${
    props.disabled
      ? `&& .check-radio-display {
        border-color: ${props.styles_colors_disabled};
      }`
      : `&:hover .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }
      &.focus .check-radio-display {
        border-color: ${props.styles_colors_primary};
      }`
  }

  .check-radio-display {
    margin-right: ${props.styles_checkRadio_labelMargin};
    display: inline-flex;
    align-items: center;
    border-radius: 1em;
    padding-top: ${props.styles_checkRadio_paddingTop};
    padding-bottom: ${props.styles_checkRadio_paddingBottom};
    padding-right: ${props.styles_checkRadio_paddingRight};
    padding-left: ${props.styles_checkRadio_paddingLeft};
    border: ${props.styles_checkRadio_borderWidth} solid
      ${props.styles_colors_secondary};
    width: calc(${props.styles_icon_width} * 2);
    height: ${props.styles_icon_width};
    background-color: ${
      props.disabled
        ? props.styles_colors_disabled
        : props.styles_colors_background
    };

    svg {
      width: ${props.styles_icon_width};
      height: ${props.styles_icon_width};
      border-radius: 100%;
      ${
        props.checked
          ? `margin-left: ${props.styles_icon_width};
          background-color: ${props.styles_colors_primary};`
          : `background-color: ${props.styles_colors_secondary};`
      }
      fill-opacity: 0;
    }
  }

  input {
    opacity: 0;
    width: 0;
  }
`
);

export { CheckBox, Radio, Switch };
