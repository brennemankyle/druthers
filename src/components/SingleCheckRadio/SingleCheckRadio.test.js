import React, { useState } from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { MockSvg } from "../../mocks";
import SingleCheckRadio from "./SingleCheckRadio";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn()
}));

const props = {
  name: "test",
  value: "value",
  className: "test",
  label: "test label",
  disabled: false,
  toggle: false,
  checked: false,
  multiple: false,
  svg_Checkmark: MockSvg,
  onChange: jest.fn(),
  onClick: jest.fn()
};

const setState = jest.fn();

beforeEach(() => {
  jest.clearAllMocks();
  useState.mockImplementation(init => [init, setState]);
});

it("renders radio", () => {
  const wrapper = shallow(<SingleCheckRadio {...props} />);

  expect(toJson(wrapper)).toMatchSnapshot();
});

it("renders checkbox", () => {
  const wrapper = shallow(<SingleCheckRadio {...props} multiple={true} />);

  expect(toJson(wrapper)).toMatchSnapshot();
});

it("renders toggle", () => {
  const wrapper = shallow(<SingleCheckRadio {...props} toggle={true} />);

  expect(toJson(wrapper)).toMatchSnapshot();
});

it("should click", () => {
  const wrapper = shallow(<SingleCheckRadio {...props} />);

  wrapper.find("input").simulate("click");

  expect(props.onClick).toBeCalled();
});

it("should change", () => {
  const wrapper = shallow(<SingleCheckRadio {...props} />);

  wrapper.find("input").simulate("change");

  expect(props.onChange).toBeCalled();
});

it("should focus", () => {
  const wrapper = shallow(<SingleCheckRadio {...props} />);

  expect(setState).not.toBeCalled();

  wrapper.find("input").simulate("focus");

  expect(setState).toBeCalledWith(true);
});

it("should blur", () => {
  const wrapper = shallow(<SingleCheckRadio {...props} />);

  expect(setState).not.toBeCalled();

  wrapper.find("input").simulate("blur");

  expect(setState).toBeCalledWith(false);
});
