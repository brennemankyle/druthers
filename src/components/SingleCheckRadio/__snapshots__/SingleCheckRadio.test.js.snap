// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders checkbox 1`] = `
<label
  className="test "
  key="value"
>
  <input
    checked={false}
    disabled={false}
    name="test"
    onBlur={[Function]}
    onChange={[MockFunction]}
    onClick={[MockFunction]}
    onFocus={[Function]}
    type="checkbox"
    value="value"
  />
  <div
    className="check-radio-display"
  >
    <MockSvg />
  </div>
  test label
</label>
`;

exports[`renders radio 1`] = `
<label
  className="test "
  key="value"
>
  <input
    checked={false}
    disabled={false}
    name="test"
    onBlur={[Function]}
    onChange={[MockFunction]}
    onClick={[MockFunction]}
    onFocus={[Function]}
    type="radio"
    value="value"
  />
  <div
    className="check-radio-display"
  >
    <MockSvg />
  </div>
  test label
</label>
`;

exports[`renders toggle 1`] = `
<label
  className="test "
  key="value"
>
  <input
    checked={false}
    disabled={false}
    name="test"
    onBlur={[Function]}
    onChange={[MockFunction]}
    onClick={[MockFunction]}
    onFocus={[Function]}
    type="checkbox"
    value="value"
  />
  <div
    className="check-radio-display"
  >
    <MockSvg />
  </div>
  test label
</label>
`;
